pkg/server/api/localauthority/v1/service.go

```diff
@@ -0,0 +1,302 @@
+	log := rpccontext.Logger(ctx)
+
+	current := s.ca.GetCurrentX509CASlot()
+	if current.Status() != journal.Status_ACTIVE {
+		return nil, api.MakeErr(log, codes.Unavailable, "server is initializing", nil)
+	}
+
+	if err := s.ca.PrepareX509CA(ctx); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to prepare X.509 authority", err)
+	}
+
+	slot := s.ca.GetNextX509CASlot()
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.PrepareX509AuthorityResponse{
+		PreparedAuthority: &localauthorityv1.AuthorityState{
+			AuthorityId: slot.AuthorityID(),
+			ExpiresAt:   slot.NotAfter().Unix(),
+		},
+	}, nil
+}
+
+func (s *Service) ActivateX509Authority(ctx context.Context, req *localauthorityv1.ActivateX509AuthorityRequest) (*localauthorityv1.ActivateX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+	if req.AuthorityId != "" {
+		log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+	}
+
+	nextSlot := s.ca.GetNextX509CASlot()
+
+	switch {
+	// Authority ID is required
+	case req.AuthorityId == "":
+		return nil, api.MakeErr(log, codes.InvalidArgument, "no authority ID provided", nil)
+
+	/// Only next local authority can be Activated
+	case req.AuthorityId != nextSlot.AuthorityID():
+		return nil, api.MakeErr(log, codes.InvalidArgument, "unexpected authority ID", nil)
+
+	// Only PREPARED local authorities can be Activated
+	case nextSlot.Status() != journal.Status_PREPARED:
+		return nil, api.MakeErr(log, codes.Internal, "only Prepared authorities can be activated", nil)
+	}
+
+	// Move next into current and reset next to clean CA
+	s.ca.RotateX509CA()
+
+	current := s.ca.GetCurrentX509CASlot()
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: current.AuthorityID(),
+		ExpiresAt:   current.NotAfter().Unix(),
+	}
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.ActivateX509AuthorityResponse{
+		ActivatedAuthority: state,
+	}, nil
+}
+
+func (s *Service) TaintX509Authority(ctx context.Context, req *localauthorityv1.TaintX509AuthorityRequest) (*localauthorityv1.TaintX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+	if req.AuthorityId != "" {
+		log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+	}
+
+	nextSlot := s.ca.GetNextX509CASlot()
+
+	switch {
+	// Authority ID is required
+	case req.AuthorityId == "":
+		return nil, api.MakeErr(log, codes.InvalidArgument, "no authority ID provided", nil)
+
+		// It is not possible to taint Active authority
+	case req.AuthorityId == s.ca.GetCurrentX509CASlot().AuthorityID():
+		return nil, api.MakeErr(log, codes.InvalidArgument, "unable to taint current local authority", nil)
+
+	// Only next local authority can be tainted
+	case req.AuthorityId != nextSlot.AuthorityID():
+		return nil, api.MakeErr(log, codes.InvalidArgument, "unexpected authority ID", nil)
```

Comment:
We only want to taint "old" authorities, right?

Reply:
yeap, the next case validates that the next is not in prepared status (so only old ones can be tainted)